File: 7_1/DemoCabinRental.py
# Розробити клас CabinRental:
# - ціле значення номера кімнати
# - дійсне значення тижневої вартості аренди кімнати
# - методи-геттери для значень полів
# - конструктор має отримувати ціле число - номер кімнати
# - вартість аренди: 950 для кімнат 1, 2, 3
#   інші кімнати - 1100

# Розробити клас-нащадок HolidayCabinRental
# - конструктор отримує номер та додає 150 до аренди

# Розробити демо програму DemoCabinRental
# - створюється об'єкт кожного класу
# - демонстрація усіх методів

# Кожен клас повинен бути збереженим у окремому файлі

class CabinRental:

    def __init__(self,cabinNumber):
        self.__cabinNumber = cabinNumber
        if (cabinNumber >= 1) and (cabinNumber <= 3):
            self.__rentalRate = 950
        else:
            self.__rentalRate = 1100

    @property
    def cabinNumber(self):
        return self.__cabinNumber

    @property
    def rentalRate(self):
        return self.__rentalRate

    def printInfo(self):
        print(f'Cabin number: {self.__cabinNumber}'
              f'\nRental rate: {self.__rentalRate}')

class HolidayCabinRental(CabinRental):

    def __init__(self,cabinNumber):
        super().__init__(cabinNumber)
        self.__rentalRate = super().rentalRate + 150

    def printInfo(self):
        print(f'Cabin number: {self.cabinNumber}'
              f'\nHoliday rental rate: {self.__rentalRate}')

c1 = CabinRental(3)
c2 = CabinRental(10)

c1.printInfo()
c2.printInfo()

print(f"Rental rate in c2 is {c2.rentalRate}")

c3 = HolidayCabinRental(3)
c4 = HolidayCabinRental(10)

print("")
c3.printInfo()
c4.printInfo()

print(f"Holiday rental rate in c4 is {c4.rentalRate}")

# ПРОБЛЕМА!
# Завдання. Подумайте і модифікуйте клас HolidayCabinRental,
# щоб вивід поточної програми повністю був аналогічний виводу
# програми DemoCabinRental.java.
# Підказка. Пошукайте, яким чином можна створювати поля, що
# є аналогами полів protected мови java.

File: 7_2/DemoCabinRental.java
// Create an abstract class named Book. Include a String field for the book’s title
// and a double field for the book’s price. Within the class, include a constructor
// that requires the book title, and add two get methods—one that returns the title
// and one that returns the price. Include an abstract method named setPrice().
// Create two child classes of Book: Fiction and NonFiction. Each must include a
// setPrice() method that sets the price for all Fiction Books to $24.99 and for
// all NonFiction Books to $37.99. Write a constructor for each subclass, and
// include a call to setPrice() within each. Write an application demonstrating
// that you can create both a Fiction and a NonFiction Book and display their
// fields. Save the files as Book.java, Fiction.java, NonFiction.java, and
// UseBook.java.

class CabinRental {
    private int cabinNumber;
    protected double rentalRate;

    CabinRental( int cabinNumber ) {
        this.cabinNumber = cabinNumber;
        if(( cabinNumber >= 1) && (cabinNumber <= 3))
            rentalRate = 950.0;
        else
            rentalRate = 1100.0;

    }

    public int get_cabinNumber() {
        return cabinNumber;
    }

    public double get_rentalRate() {
        return rentalRate;
    }

    public void printInfo() {
        System.out.println("Cabin number: " + cabinNumber +
                           "\nRental rate: " + rentalRate);
    }
}

class HolidayCabinRental extends CabinRental {

    HolidayCabinRental(int cabinNumber) {
        super(cabinNumber);
        rentalRate += 150.0;
    }

    public void printInfo() {
        System.out.println("Cabin number: " + get_cabinNumber() +
                           "\nHoliday rental rate: " + rentalRate);
    }
}

class DemoCabinRental {

    public static void main(String[] args) {

        CabinRental c1 = new CabinRental(3);
        CabinRental c2 = new CabinRental(10);

        c1.printInfo();
        c2.printInfo();

        System.out.println("Rental rate in c2 is " +
            c2.get_rentalRate());

        HolidayCabinRental c3 = new HolidayCabinRental(3);
        HolidayCabinRental c4 = new HolidayCabinRental(10);

        System.out.println("");
        c3.printInfo();
        c4.printInfo();

        System.out.println("Holiday rental rate in c4 is " +
            c4.get_rentalRate());
    }
}
